custom/shared/hooks/useSharedState.js

```diff
@@ -0,0 +1,112 @@
+import React, { useState, useEffect, useCallback, useRef } from 'react';
+import { useCallState } from '../contexts/CallProvider';
+
+// @params initialValues - initial values of the shared state.
+// @params broadcast - share the state with the other participants whenever the state changes.
+export const useSharedState = ({ initialValues = {}, broadcast = true }) => {
+  const { callObject } = useCallState();
+  const stateRef = useRef(null);
+
+  const [state, setState] = useState({
+    sharedState: initialValues,
+    setAt: undefined,
+  });
+
+  // handling the app-message event, to check if the state is being shared.
+  const handleAppMessage = useCallback(
+    event => {
+      // two types of events -
+      // 1. Request shared state (request-shared-state)
+      // 2. Set shared state (set-shared-state)
+      switch (event.data?.message?.type) {
+        // if we receive a request-shared-state message type, we check if the user has any previous state,
+        // if yes, we will send the shared-state to everyone in the call.
+        case 'request-shared-state':
+          if (!stateRef.current.setAt) return;
+          // random delay to avoid too many messages being sent at the same time from different clients.
+          const randomDelay = 200 + Math.ceil(1000 * Math.random());
+          setTimeout(() => {
+            callObject.sendAppMessage(
+              {
+                message: {
+                  type: 'set-shared-state',
+                  value: stateRef.current,
+                },
+              },
+              '*',
+            );
+          }, randomDelay);
+          break;
+        // if we receive a set-shared-state message type then, we check the state timestamp with the local one and
+        // we set the latest shared-state values into the local state.
+        case 'set-shared-state':
+          if (
+            stateRef.current.setAt &&
+            stateRef.current.setAt > event.data.message.value.setAt
+          )
+            return;
+          setState(event.data.message.value);
+          break;
+      }
+    },
+    [stateRef, callObject],
+  );
+
+  // whenever local user joins, we request the state from everyone in the call.
```

Comment:
We should only request state from one remote participant, who joined previously.
Requesting state from all other participants will likely lead to message bottlenecks.

I suggest to change `handleJoinedMeeting` to:
- request state from one remote participant
- filter remote participants where `joined_at` is smaller than the local participant

Example (untested code from memory):
```js
const localParticipant = callObject.participants().local;
const remoteParticipants = Object.values(callObject.participants()).filter((p) => !p.local && new Date(p.joined_at) < new Date(localParticipant.joined_at));
const randomPeer = Math.floor(Math.random() * (remoteParticipants.length - 1));
```
- set the interval to a randomized value